sdk/java-sdk/src/main/java/kalix/javasdk/action/Action.java

```diff
@@ -24,6 +24,7 @@
 import kalix.javasdk.impl.action.ActionEffectImpl;
 import kalix.javasdk.timer.TimerScheduler;
 import kalix.javasdk.impl.timer.TimerSchedulerImpl;
+import reactor.core.publisher.Mono;
```

Comment:
Mono implements `org.reactivestream.Publisher` so we could instead on Reactive Streams. 

There is a potential issue though, `Flux` also implements `Publisher`. So one could pass a `Flux` to `asyncReply`. We need to check how that will work. 

What we will happen if we have `Flux[T] => CompletionState[T]`? I guess it will only be able to emit the first element and then it will close the stream.

Reply:
Actually, letting the javasdk depend on Reactor is not good. 
We either depend on Reactive Streams or we only offer CompletionStage.

I think we can accept `Publisher` and document that only the first element will be consumed.

Reply:
I would agree with accepting `Publisher` and documenting it. The way I'm thinking about it is that you get back a lazy-List-like instance (the `Publisher`). I think it's intuitive to find out you can get one or more elements only at runtime. Nobody has a problem reasoning about lists.

Reply:
Thanks for the feedback, updated now using reactive streams. I wonder if we should also support the [Flow.Publisher](https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.Publisher.html) from java..

> The way I'm thinking about it is that you get back a lazy-List-like instance (the `Publisher`).
Not sure if I followed your suggestion.. current approach only takes the first element and ignores the rest. Let me know if your suggestion is different.

Reply:
Looking at the implementation now I think I understand better what @octonato was saying. Which makes me think exactly what you are pointing at Eduardo.

> I wonder if we should also support the [Flow.Publisher](https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.Publisher.html) from java.

From the beginning, I thought we were going to support `Flow` and `Mono` (through `Publisher`), document it and let the user deal with one or multiple elements.  I see now that I misinterpreted the previous [comment](https://github.com/lightbend/kalix-jvm-sdk/pull/1426#discussion_r1091217323).

Reply:
> I thought we were going to support `Flow` and `Mono` (through `Publisher`)

well, we do support because both implement Publisher. So user can actually pass in a Flow. But in that case we'll only pick the first element to build the reply (I will add a test btw).

What I was mentioning as well is support for `Flow.Publisher` which is the semantically-equivalent native implementation of the reactive streams directly on the jdk, since eventually all libraries should move to use those.

Reply:
I think accepting Publisher is a bad idea, the SPI interface is to allow a stream of async messages but in this case what we really want to support is specifically a Mono, not arbitrary streaming publishers. Better keep the API tight and explicit about what it supports IMO.

Reply:
Yeah, I think you are right. The initial idea was to avoid depending on Reactor directly and to avoid writing wrappers around the Java SDK. 

But your point that `Publisher` semantics is about stream and not about a single value is solid.  

I don't think it's worth to build a wrapper around the java sdk just for that. Reactor users can still build a `CompletionStage` from a `Mono`. And I don't think we should add a direct dependency on Project Reactor either. 

If we have other reasons to build a wrapper, then we can re-evaluate. 

Today, the reasons for the Spring wrapper are:

1. support for async calls accepting `Mono` 
2. error codes that don't expose gprc libraries. 
3. typed events - `EventSourcedEntity<State, Event>`

To avoid the wrapper, we can:

1. not support `Mono` directly, users will need to convert to `CompletionStage` 
2. introduce Kalix StatusCodes so code-first users don't need to deal with gRPC codes
3. expand EventSourcedEntity to have another typed parameter. Codegen can set it to `Object` or we can introduced a new type like in `EventSourcedEntity<State> extends TypedEventSourcedEntity<State, Object>`. 

So, in retrospect, we have acceptable solution for all three reasons and therefore I'm less inclined to build a wrapper.